A31-xds-routeaction-support.md

```diff
@@ -0,0 +1,184 @@
+implement.
+
+
+## Proposal
+
+This proposal is broken up into two sections: the addition of new functionality
+to gRPC itself, and then a description of using that functionality to implement
+support for xDS RouteActions.
+
+### New Functionality in gRPC
+
+In order to allow the xDS matchers to configure gRPC behavior, but to avoid
+implementing support for them directly in gRPC, a new "per-RPC Config Selector"
+will be added for internal-only use.  This will be provided by the Name Resolver
+component when calling `Update` on the channel in a way that does not extend our
+publicly-visible API, similarly to how the xDS Client is provided to the Load
+Balancing Policy.
+
+If present, this Config Selector will be invoked once at the beginning of every
+RPC (after the name resolver has produced a result), and it will produce four
+things:
+
+1.  The same per-RPC configuration which is currently determined by the Service
+    Config's MethodConfig.  A default implementation will return the Service
+    Config's selected values.
+1.  An optional interceptor or callback.  The interceptor would be invoked by
+    the channel immediately for the RPC, or the callback would be invoked at
+    some point after the channel knows the RPC will no longer call the Load
+    Balancing Policy's Pick method and before the end of the RPC.  This is
+    necessary for cluster lifecycle management, which will be explained in
+    detail later.
+1.  Information which will be passed to the Load Balancing Policy's Pick method
+    by the channel.  E.g. in Go this could be done via the Context, and Java may
+    use CallOptions.  If an interceptor is used for (2) it is possible that the
+    interceptor may be capable of satisfying this requirement directly.
+1.  An RPC status; if not OK, the RPC is terminated with this status.  This is
+    needed in case the RPC cannot be routed.  Like (3), this could be satisfied
+    via an interceptor.
+
+Note that, for reasons described below, the Load Balancing Policy's Update
+method must be called and return before the Config Selector may be used by the
+channel.
+
+
+### Implementation for xDS RouteAction Support
+
+The "xds" resolver will return a Config Selector which applies the xDS matchers
+to the RPC (using the Method name and RPC Header data provided to it) and uses
+those inputs to perform the initial routing steps for the RPC based upon the RDS
+Routes.  This will enable it to set the timeouts and other settings
+appropriately in the MethodConfig according to the RouteActions.  Since the
+routing decisions are made here, it will also return the cluster name to the
+top-level Load Balancing Policy in order to inform the policy which endpoint
+picker to use for the RPC.  Note that the Config Selector will also be assuming
+the responsibility of the current `weighted_target` Load Balancing Policy.  This
+means two things: 1. connectivity state is not considered when making this
```

Comment:
We should explicitly note that this is a change from the behavior specified in [gRFC A28](https://github.com/grpc/proposal/blob/master/A28-xds-traffic-splitting-and-routing.md), where we explicitly stated that we'd do this differently from Envoy.